File: dbusmenu4j/src/main/java/io/gitlab/jfronny/dbusmenu4j/SwingMenu.java
package io.gitlab.jfronny.dbusmenu4j;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.imageio.ImageIO;
import javax.swing.*;
import java.awt.event.*;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.lang.reflect.Modifier;
import java.util.*;
import java.util.stream.Collectors;

public class SwingMenu<R extends SwingRootMenu<R, ?>, T extends SwingMenuHolder<R>> extends Menu.Abstract {
    private static final Logger log = LoggerFactory.getLogger("dbusmenu4j/SwingMenu");

    protected final @Nullable JMenuItem menuItem;
    private final int id;
    protected final T holder;

    public SwingMenu(@Nullable JMenuItem menuItem, int id, T holder) {
        this.menuItem = menuItem;
        this.id = id;
        this.holder = holder;
    }

    public SwingMenu(JMenuItem menuItem, T holder) {
        this(menuItem, holder.getId(menuItem), holder);
    }

    public SwingMenu(int id, T holder) {
        this(null, id, holder);
    }

    @Override
    public int getId() {
        return id;
    }

    @Override
    public boolean isSeparator() {
        return menuItem == null;
    }

    @Override
    public @NotNull String getLabel() {
        if (menuItem == null) return "";
        String text = menuItem.getText();
        return text == null ? "" : text;
    }

    @Override
    public boolean isEnabled() {
        if (menuItem == null) return false;
        return menuItem.isEnabled();
    }

    @Override
    public boolean isVisible() {
        if (menuItem == null) return false;
        return menuItem.isVisible();
    }

    private final Object iconDataLock = new Object();
    private boolean hasIconData = false;
    private byte[] iconData = null;

    @Override
    public byte @Nullable [] getIconData() {
        if (hasIconData) return iconData;
        synchronized (iconDataLock) {
            if (hasIconData) return iconData;
            hasIconData = true;
            if (menuItem == null) return null;
            Icon icon = menuItem.getIcon();
            if (icon == null) return null;
            // This is somewhat expensive, but we need to do it
            // At least avoid doing it multiple times
            int width = icon.getIconWidth();
            int height = icon.getIconHeight();

            BufferedImage bufferedImage = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);

            var graphics = bufferedImage.createGraphics();
            icon.paintIcon(menuItem, graphics, 0, 0);
            graphics.dispose();

            try (ByteArrayOutputStream stream = new ByteArrayOutputStream()) {
                ImageIO.write(bufferedImage, "png", stream);
                return iconData = stream.toByteArray();
            } catch (Exception e) {
                log.error("Failed to convert icon to byte array", e);
                return iconData = null;
            }
        }
    }

    private static final Map<Integer, String> keyEvents = Arrays.stream(KeyEvent.class.getFields())
            .filter(it -> it.getModifiers() == (Modifier.PUBLIC | Modifier.STATIC | Modifier.FINAL) )
            .filter(it -> it.getName().startsWith("VK_"))
            .collect(Collectors.toMap(
                    s -> {
                        try {
                            return s.getInt(null);
                        } catch (IllegalAccessException e) {
                            throw new RuntimeException(e);
                        }
                    },
                    s -> s.getName().substring(3),
                    (a, b) -> a
            ));

    @Override
    public String @Nullable [] getShortcut() {
        if (menuItem == null) return null;
        KeyStroke ks = menuItem.getAccelerator();
        if (ks == null) return null;
        List<String> result = new ArrayList<>();
        int modifiers = ks.getModifiers();
        if ((modifiers & InputEvent.SHIFT_DOWN_MASK) != 0) result.add("Shift");
        if ((modifiers & InputEvent.CTRL_DOWN_MASK) != 0) result.add("Ctrl");
        if ((modifiers & InputEvent.META_DOWN_MASK) != 0) result.add("Meta");
        if ((modifiers & InputEvent.ALT_DOWN_MASK) != 0) result.add("Alt");
        if ((modifiers & InputEvent.ALT_GRAPH_DOWN_MASK) != 0) result.add("AltGraph");
        if ((modifiers & InputEvent.BUTTON1_DOWN_MASK) != 0) result.add("Button1");
        if ((modifiers & InputEvent.BUTTON2_DOWN_MASK) != 0) result.add("Button2");
        if ((modifiers & InputEvent.BUTTON3_DOWN_MASK) != 0) result.add("Button3");
        String keyEvent = keyEvents.get(ks.getKeyCode());
        result.add(keyEvent != null ? keyEvent : "UNKNOWN");
        return result.toArray(String[]::new);
    }

    @Override
    public @Nullable String getToggleType() {
        return switch (menuItem) {
            case JRadioButtonMenuItem b -> "radio";
            case JCheckBoxMenuItem c -> "checkmark";
            case null, default -> null;
        };
    }

    @Override
    public int getToggleState() {
        String toggleType = getToggleType();
        if (toggleType == null || toggleType.isEmpty()) return -1;
        if (menuItem.isSelected()) return 1;
        return 0;
    }

    private List<Menu> children = null;
    @Override
    public @Nullable List<Menu> getChildren() {
        return children;
    }

    @Override
    public void onEvent() {
        switch (menuItem) {
            case JCheckBoxMenuItem c -> c.setSelected(!c.isSelected());
            case JRadioButtonMenuItem r -> r.setSelected(true);
            case null, default -> {}
        }
        var event = new ActionEvent(menuItem, ActionEvent.ACTION_PERFORMED, menuItem.getActionCommand());
        for (ActionListener listener : menuItem.getActionListeners()) {
            listener.actionPerformed(event);
        }
    }

    public void syncChildren(int depth) {
        if (menuItem instanceof JMenu menu) {
            children = new ArrayList<>();
            int len = menu.getItemCount();
            for (int i = 0; i < len; i++) {
                JMenuItem item = menu.getItem(i);
                if (item == null) continue;
                SwingMenu<R, T> swingMenu = child(item, holder);
                if (depth > 0) swingMenu.syncChildren(depth - 1);
                children.add(swingMenu);
            }
        }
    }

    protected SwingMenu<R, T> child(JMenuItem item, T holder) {
        return new SwingMenu<>(item, holder);
    }

    @Override
    public String toString() {
        return "SwingMenu(id=" + id + ", menuItem=" + menuItem + ")";
    }
}

File: src/main/kotlin/io/gitlab/jfronny/globalmenu/proxy/SwingMenu.kt
package io.gitlab.jfronny.globalmenu.proxy

import com.intellij.openapi.actionSystem.impl.ActionMenu
import com.intellij.openapi.actionSystem.impl.ActionMenuItem
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.EDT
import io.gitlab.jfronny.dbusmenu4j.Menu
import io.gitlab.jfronny.globalmenu.GlobalMenu
import io.gitlab.jfronny.globalmenu.buildArray
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.apache.commons.io.output.ByteArrayOutputStream
import java.awt.event.InputEvent
import java.awt.event.KeyEvent
import java.awt.image.BufferedImage
import java.lang.reflect.Modifier
import javax.imageio.ImageIO
import javax.swing.*

class SwingMenu(private val menuItem: JMenuItem?, id: Int, private val holder: SwingMenuHolder) : Menu.Abstract() {
    private val id_ = id;
    override fun getId() = id_
    constructor(menuItem: JMenuItem, holder: SwingMenuHolder) : this(menuItem, holder.getId(menuItem), holder)
    constructor(id: Int, holder: SwingMenuHolder) : this(null, id, holder)

    override fun isSeparator() = menuItem == null
    override fun getLabel() = menuItem?.text ?: ""
    override fun isEnabled() = menuItem?.isEnabled ?: false
    override fun isVisible() = menuItem?.isVisible ?: false
    private val iconData_: ByteArray? by lazy { menuItem?.icon?.let { icon ->
        // This is somewhat expensive, but we need to do it
        // At least avoid doing it multiple times
        val width = icon.iconWidth
        val height = icon.iconHeight

        val bufferedImage = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)

        bufferedImage.createGraphics().apply {
            icon.paintIcon(menuItem, this, 0, 0)
            dispose()
        }

        try {
            ByteArrayOutputStream().use { stream ->
                ImageIO.write(bufferedImage, "png", stream)
                stream.toByteArray()
            }
        } catch (e: Exception) {
            GlobalMenu.Log.error("Failed to convert icon to byte array", e)
            null
        }
    } }
    override fun getIconData() = iconData_

    override fun getShortcut() = menuItem?.accelerator?.let { ks ->
        buildArray {
            val modifiers = ks.modifiers
            if (modifiers and InputEvent.SHIFT_DOWN_MASK != 0) accept("Shift")
            if (modifiers and InputEvent.CTRL_DOWN_MASK != 0) accept("Ctrl")
            if (modifiers and InputEvent.META_DOWN_MASK != 0) accept("Meta")
            if (modifiers and InputEvent.ALT_DOWN_MASK != 0) accept("Alt")
            if (modifiers and InputEvent.ALT_GRAPH_DOWN_MASK != 0) accept("AltGraph")
            if (modifiers and InputEvent.BUTTON1_DOWN_MASK != 0) accept("Button1")
            if (modifiers and InputEvent.BUTTON2_DOWN_MASK != 0) accept("Button2")
            if (modifiers and InputEvent.BUTTON3_DOWN_MASK != 0) accept("Button3")
            accept(keyEvents[ks.keyCode] ?: "UNKNOWN")
        }
    }

    override fun getToggleType() = when (menuItem) {
        is ActionMenuItem -> {
            //TODO handle action items
            if (menuItem.isToggleable) "checkmark"
            else null
        }
        is JRadioButtonMenuItem -> "radio"
        is JCheckBoxMenuItem -> "checkmark"
        else -> null
    }
    override fun getToggleState() = if (toggleType?.isNotEmpty() == true) if (menuItem!!.isSelected) 1 else 0 else -1
    private var _children: List<Menu>? = null
    override fun getChildren() = _children

    override fun onEvent() {
        if (menuItem == null) return
        ApplicationManager.getApplication().invokeLater(menuItem::doClick)
        when (menuItem) {
            is ActionMenuItem -> {}
            is ActionMenu -> {}
            is JCheckBoxMenuItem -> menuItem.isSelected = !menuItem.isSelected
            is JRadioButtonMenuItem -> menuItem.isSelected = true
        }
    }

    override fun update() {
        super.update()
        try {
            if (menuItem is ActionMenu) {
                runBlocking {
                    launch(Dispatchers.EDT) {
                        menuItem.removeAll()
                        menuItem.isSelected = true
                        menuItem.fillMenu()
                        syncChildren(2)
                    }
                }
            }
        } catch (e: Exception) {
            GlobalMenu.Log.error("Failed to update menu", e)
            throw e
        }
    }

    fun syncChildren(deepness: Int) {
//        GlobalMenu.Log.warn("Syncing children for $label")
        _children = when (menuItem) {
            is ActionMenu -> {
                val ch = mutableListOf<Menu>()
                for (each in menuItem.popupMenu.components) {
                    if (each == null) continue
                    if (each is JSeparator) {
                        ch.add(SwingMenu(System.identityHashCode(each), holder))
                        continue
                    }
                    if (each !is JMenuItem) {
                        continue
                    }
                    val cmi = SwingMenu(each, holder)
                    if (deepness > 1) {
                        if (each is ActionMenu) {
                            each.removeAll()
                            each.isSelected = true
                            each.fillMenu() // This is REALLY expensive and on the EDT TODO: find out whether we can avoid it
                        }
                        cmi.syncChildren(deepness - 1)
                    }
                    ch.add(cmi)
                }
                ch
            }
            is JMenu -> (0 until menuItem.itemCount).map { SwingMenu(menuItem.getItem(it), holder) }
            else -> null
        }
    }

    override fun toString(): String {
        return "SwingMenu(id=$id_, menuItem=$menuItem)"
    }

    companion object {
        val keyEvents: Map<Int, String> = KeyEvent::class.java.fields
            .filter { it.modifiers == Modifier.PUBLIC or Modifier.STATIC or Modifier.FINAL && it.name.startsWith("VK_") }
            .associate { it.getInt(null) to it.name.substring(3) }
    }
}